src/conn/conn_dhandle.c

```diff
@@ -708,6 +708,15 @@ __wt_conn_dhandle_discard(WT_SESSION_IMPL *session)
 	__wt_session_close_cache(session);
 	F_SET(session, WT_SESSION_NO_DATA_HANDLES);
 
+	/*
+	 * The connection may have an open metadata cursor handle. We cannot
+	 * close it before now because it's potentially used when discarding
+	 * other open data handles. Close it before discarding the underlying
+	 * metadata handle.
+	 */
+	if (session->meta_cursor != NULL)
+		WT_TRET(session->meta_cursor->close(session->meta_cursor));
+
```

Comment:
On second thought, does it make sense to move this into `__wt_session_close_cache`?